### PR TITLE
Revalidate transaction pool off of ledger state when updating best tip

### LINKED ISSUE #10879: berkeley-qa: Crashed Mina Daemon With Multiple ZkApp Tx and User Tx
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/MinaProtocol/mina/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report. Please post those things in GitHub Discussions: https://github.com/MinaProtocol/mina/discussions

### Description

I was sending multiple zkapp transactions along with many user transactions and crashed the daemon:

```	error: "Tried to handle a committed transaction in the pool but its nonce doesn't match the head of the queue for that sender"
2022-05-06 16:12:57 UTC [Fatal] Unhandled top-level exception: $exn
Generating crash report
	exn: {
  "commit_id": "552f5479ccccc7b501a84a843024bf62bb845d4a",
  "sexp": [
    "monitor.ml.Error",
    [
      "Failure",
      "Tried to handle a committed transaction in the pool but its nonce doesn't match the head of the queue for that sender"
    ],
    [
      "Raised at Stdlib.failwith in file \"stdlib.ml\", line 29, characters 17-33",
      "Called from Network_pool__Transaction_pool.Make0.Resource_pool.handle_transition_frontier_diff.(fun) in file \"src/lib/network_pool/transaction_pool.ml\", line 666, characters 18-36",
      "Called from Stdlib__list.fold_left in file \"list.ml\", line 121, characters 24-34",
      "Called from Base__List0.fold in file \"src/list0.ml\" (inlined), line 21, characters 22-52",
      "Called from Network_pool__Transaction_pool.Make0.Resource_pool.handle_transition_frontier_diff in file \"src/lib/network_pool/transaction_pool.ml\", line 605, characters 8-1023",
      "Called from Base__Result.try_with in file \"src/result.ml\", line 179, characters 9-15",
      "Caught by monitor coda"
    ]
  ],
  "backtrace": [
    "Raised at Stdlib.failwith in file \"stdlib.ml\", line 29, characters 17-33",
    "Called from O1trace.exec_thread in file \"src/lib/o1trace/o1trace.ml\", line 68, characters 6-27",
    "Called from Pipe_lib__Strict_pipe.Reader0.Merge.iter.read_deferred.(fun) in file \"src/lib/pipe_lib/strict_pipe.ml\", line 162, characters 30-39",
    "Called from Async_kernel__Deferred0.bind.(fun) in file \"src/deferred0.ml\", line 54, characters 64-69",
    "Called from Async_kernel__Job_queue.run_job in file \"src/job_queue.ml\" (inlined), line 128, characters 2-5",
    "Called from Async_kernel__Job_queue.run_jobs in file \"src/job_queue.ml\", line 169, characters 6-47"
  ]
}
2022-05-06 16:12:57 UTC [Fatal] Unhandled top-level exception: $exn
Generating crash report
	exn: {
  "commit_id": "552f5479ccccc7b501a84a843024bf62bb845d4a",
  "sexp": [
    "monitor.ml.Error",
    [
      "Failure",
      "timing task `processing_transaction_pool_transition_frontier_diffs` failed, exception reported to parent monitor"
    ],
    [
      "Raised at Stdlib.failwith in file \"stdlib.ml\", line 29, characters 17-33",
      "Called from O1trace.exec_thread in file \"src/lib/o1trace/o1trace.ml\", line 68, characters 6-27",
      "Called from Pipe_lib__Strict_pipe.Reader0.Merge.iter.read_deferred.(fun) in file \"src/lib/pipe_lib/strict_pipe.ml\", line 162, characters 30-39",
      "Called from Async_kernel__Deferred0.bind.(fun) in file \"src/deferred0.ml\", line 54, characters 64-69",
      "Called from Async_kernel__Job_queue.run_job in file \"src/job_queue.ml\" (inlined), line 128, characters 2-5",
      "Called from Async_kernel__Job_queue.run_jobs in file \"src/job_queue.ml\", line 169, characters 6-47",
      "Caught by monitor coda"
    ]
  ],
  "backtrace": [
    "Raised at Stdlib.failwith in file \"stdlib.ml\", line 29, characters 17-33",
    "Called from O1trace.exec_thread in file \"src/lib/o1trace/o1trace.ml\", line 68, characters 6-27",
    "Called from Pipe_lib__Strict_pipe.Reader0.Merge.iter.read_deferred.(fun) in file \"src/lib/pipe_lib/strict_pipe.ml\", line 162, characters 30-39",
    "Called from Async_kernel__Deferred0.bind.(fun) in file \"src/deferred0.ml\", line 54, characters 64-69",
    "Called from Async_kernel__Job_queue.run_job in file \"src/job_queue.ml\" (inlined), line 128, characters 2-5",
    "Called from Async_kernel__Job_queue.run_jobs in file \"src/job_queue.ml\", line 169, characters 6-47"
  ]
}
2022-05-06 16:12:57 UTC [Fatal] libp2p_helper process died unexpectedly: "died after receiving sigkill (signal number 9)"
2022-05-06 16:12:57 UTC [Error] Encountered $error while asking libp2p_helper for peers
	error: {
  "commit_id": "552f5479ccccc7b501a84a843024bf62bb845d4a",
  "string": "libp2p_helper process died before answering"
}
2022-05-06 16:12:57 UTC [Error] Encountered $error while asking libp2p_helper for peers
	error: {
  "commit_id": "552f5479ccccc7b501a84a843024bf62bb845d4a",
  "string": "libp2p_helper process died before answering"
}
2022-05-06 16:12:57 UTC [Fatal] Unhandled top-level exception: $exn
Generating crash report
	exn: {
  "commit_id": "552f5479ccccc7b501a84a843024bf62bb845d4a",
  "sexp": [
    "monitor.ml.Error",
    [ "Mina_net2__Libp2p_helper.Libp2p_helper_died_unexpectedly" ],
    [
      "Raised at Mina_net2__Libp2p_helper.handle_libp2p_helper_termination in file \"src/lib/mina_net2/libp2p_helper.ml\", line 144, characters 8-45",
      "Called from Async_kernel__Deferred0.bind.(fun) in file \"src/deferred0.ml\", line 54, characters 64-69",
      "Called from Async_kernel__Job_queue.run_job in file \"src/job_queue.ml\" (inlined), line 128, characters 2-5",
      "Called from Async_kernel__Job_queue.run_jobs in file \"src/job_queue.ml\", line 169, characters 6-47",
      "Caught by monitor at file \"src/lib/gossip_net/libp2p.ml\", line 234, characters 31-31"
    ]
  ],
  "backtrace": [
    "Raised at Gossip_net__Libp2p.Make.T.create_libp2p.handle_mina_net2_exception in file \"src/lib/gossip_net/libp2p.ml\", line 218, characters 12-21",
    "Called from Base__Result.try_with in file \"src/result.ml\", line 179, characters 9-15"
  ]
}
2022-05-06 16:12:57 UTC [Error] Encountered $error while asking libp2p_helper for peers
	error: {
  "commit_id": "552f5479ccccc7b501a84a843024bf62bb845d4a",
  "string": "helper process already exited (doing RPC ListPeers)"
}
2022-05-06 16:14:00 UTC [Error] Encountered $error while asking libp2p_helper for peers
	error: {
  "commit_id": "552f5479ccccc7b501a84a843024bf62bb845d4a",
  "string": "helper process already exited (doing RPC ListPeers)"
}
2022-05-06 16:14:00 UTC [Error] Encountered $error while asking libp2p_helper for peers
	error: {
  "commit_id": "552f5479ccccc7b501a84a843024bf62bb845d4a",
  "string": "helper process already exited (doing RPC ListPeers)"
}
2022-05-06 16:14:29 UTC [Error] Encountered $error while asking libp2p_helper for peers
	error: {
  "commit_id": "552f5479ccccc7b501a84a843024bf62bb845d4a",
  "string": "helper process already exited (doing RPC ListPeers)"
}
2022-05-06 16:14:29 UTC [Info] Running async shutdown handler: "Close transition frontier, if exists"
2022-05-06 16:14:29 UTC [Info] Running async shutdown handler: "Remove daemon lockfile"


  ☠  Mina Daemon crashed.
   The Mina Protocol developers would like to know why!

    Please:
      Open an issue:
        <https://github.com/MinaProtocol/mina/issues/new>

      Briefly describe what you were doing and attach the crash report /root/.mina-config/coda_crash_report_2022-05-06_16-14-00.286842.tar.gz

    
```

### Steps to Reproduce

1. Start Daemon and Sync
2. Send a large number of zkapps and user transactions
3. Observe crash
...

### Expected Result

No Daemon Crash

### Actual Result

Daemon Crash

### How frequently do you see this issue?

Rarely

### What is the impact of this issue on your ability to run a node?

Blocker

### Status

```shell
N/A Mina Daemon Crashed
```

### Additional information

Messaging Deepthi on Discord with the Logs

### PR DESCRIPTION
**Pull request by nholland94**
Closes #10879.

This PR addresses bugs in which the application of parties transactions can invalidate transactions enqueued by other fee payers within the transaction pool. To solve the bug, I took the approach to just simplify the transaction pool by having it perform revalidation on the pool whenever the frontier's best tip changes, instead of incrementally updating the pool. There is an increased performance cost to doing this, but we now have a higher degree of confidence that the implementation will correctly prune invalid transactions.

The increased performance cost comes from the fact that we now need to load account states from the ledger in order to process best tip diffs from the frontier, which was previously not necessary. However, since the accounts we request from the best tip ledger are accounts which have been interacted with from the transactions contained in the chain reorg, we they should be available from "nearby" ledger masks in most cases, removing database interactions in naive cases. Additionally, the new implementation operates in `O(n)` in terms of number of accounts accessed and not `O(n)` in terms of number of transactions in the reorg, which will also show some improvement in performance.

One outstanding question about the PR: many of the accounts requested when processing a reorg were already requested from the ledger when the staged ledger diff was applied (in the naive case of a chain extension, this will be entire set of accounts we need). It is possible to try and share data between these 2 parts of the code, but this requires additional lift and will make things a bit ugly at the frontier level. Still, this may be a optimization worth visiting before we merge this.

### CODE REVIEW

File: src/lib/network_pool/indexed_pool.ml [hunk 1]
```diff
@@ -210,6 +211,17 @@ module For_tests = struct
           Transaction_hash.User_command_with_valid_signature.command applicable
         in
         check_consistent applicable ;
+        (*
```
Comment (nholland94) [inline src/lib/network_pool/indexed_pool.ml:214]: Delete.

File: src/lib/network_pool/indexed_pool.ml [hunk 2]
```diff
@@ -537,6 +549,74 @@ module Update = struct
...
+          match F_sequence.uncons sender_queue' with
+          | Some (next_txn, _) ->
+              ( insert_applicable applicable_by_fee' next_txn
+              , Map.set all_by_sender ~key:sender ~data:sender_queue' )
+          | None ->
+              (applicable_by_fee', Map.remove all_by_sender sender)
+        in
+        Some (txn, (applicable_by_fee'', all_by_sender')) )
+
+(*
```
Comment (nholland94) [inline src/lib/network_pool/indexed_pool.ml:602]: Delete.

**Review by ghost-not-in-the-shell — APPROVED** [review]

File: src/lib/network_pool/transaction_pool.ml [hunk 3]
```diff
@@ -601,89 +545,95 @@ struct
...
+                         @@ Currency.Balance.to_uint64 account.balance ) )
+          in
+          fun id ->
+            match Map.find existing_account_states_by_id id with
+            | Some state ->
+                state
             | None ->
-                ()
-            | Some time_added ->
-                [%log' info t.logger]
```
Comment (deepthiskumar) [inline src/lib/network_pool/transaction_pool.ml:634]: Add this log back?

File: src/lib/network_pool/indexed_pool.ml [hunk 4]
```diff
@@ -537,6 +538,56 @@ module Update = struct
...
+        let applicable_by_fee'', all_by_sender' =
+          let sender =
+            User_command.fee_payer
+            @@ Transaction_hash.User_command_with_valid_signature.command txn
+          in
+          let sender_queue = Map.find_exn all_by_sender sender in
+          let head_txn, sender_queue' =
+            Option.value_exn (F_sequence.uncons sender_queue)
+          in
+          assert (
```
[the remark this answers is not in the document]
Comment (nholland94) [reply src/lib/network_pool/indexed_pool.ml:578]: I'm not sure what the right thing would be to do in this case. This means the data structure's indices are malformed, so we don't really have a way to iterate over the pool in the order we need to without reconstructing the entire pool.

**Review by deepthiskumar — APPROVED** [review]

**Pull request merged**